src/parser.ts

```diff
@@ -0,0 +1,261 @@
+// TODO: Move to easyscore.ts
+type RuleFunction = () => Rule;
+type TriggerFunction = (state?: Result) => void;
+interface Rule {
+  // Lexer Rules
+  token?: string; // The token property is a string that is compiled into a RegExp.
+  noSpace?: boolean; // TODO: None of the EasyScore rules specify noSpace, so it is not used anywhere.
+
+  // Parser Rules
+  expect?: RuleFunction[];
+  zeroOrMore?: boolean;
+  oneOrMore?: boolean;
+  maybe?: boolean;
+  or?: boolean;
+  run?: TriggerFunction;
+}
+// END: Move to easyscore.ts
+//////////////////////////////////////////////////////////////////////////////////////////////////
+
+// To enable logging for this class. Set `Vex.Flow.Parser.DEBUG` to `true`.
+function L(...args: any[]) {
+  if (Parser.DEBUG) Vex.L('Vex.Flow.Parser', args);
+}
+
+export const X = Vex.MakeException('ParserError');
+
+const NO_ERROR_POS = -1;
+
+interface Result {
+  success: boolean;
+
+  // Lexer Results
+  pos?: number;
+  incrementPos?: number;
+  matchedString?: string;
+
+  // Parser Results
+  matches?: Match[];
+  numMatches?: number;
+  results?: Result[];
+  errorPos?: number; // Set to NO_ERROR if successful. N if there is an error in the string.
+}
+
+// Converts parser results into an easy to reference list that can be used in triggers.
+// eventually returns: string (including empty strings) | nested array | null
+type Match = null | string | Array<Match>;
+function flattenMatches(r: Result | Result[]): Match {
+  if ('matchedString' in r) return r.matchedString as string; // string
+  if ('results' in r) return flattenMatches(r.results as Result[]);
+  const results = r as Result[];
+  if (results.length === 1) return flattenMatches(results[0]);
+  if (results.length === 0) return null; // null
+  return results.map(flattenMatches); // nested array
+}
+
+// This is the base parser class. Given an arbitrary context-free grammar, it
+// can parse any line and execute code when specific rules are met (e.g.,
+// when a string is terminated.)
+export class Parser {
+  public static DEBUG: boolean = false;
+
+  private grammar: Grammar;
+  private line: string;
+  private pos: number;
+  private errorPos: number;
+
+  // For an example of a simple grammar, take a look at tests/parser_tests.js or
+  // the EasyScore grammar in easyscore.js.
+  constructor(grammar: Grammar) {
+    this.grammar = grammar;
+    this.line = '';
+    this.pos = 0;
+    this.errorPos = NO_ERROR_POS;
+  }
+
+  // Parse `line` using current grammar. Returns {success: true} if the
+  // line parsed correctly, otherwise returns `{success: false, errorPos: N}`
+  // where `errorPos` is the location of the error in the string.
+  public parse(line: string): Result {
+    this.line = line;
+    this.pos = 0;
+    this.errorPos = NO_ERROR_POS;
+    const result = this.expect(this.grammar.begin());
+    result.errorPos = this.errorPos;
+    return result;
+  }
+
+  private matchFail(returnPos: number): void {
```

Comment:
We are not declaring visibility of methods leaving all of them public. This is better but it is difficult to know who is using what.
@0xfe what do you think we should do going forward?

Reply:
OK, I'm happy to not declare visibility for methods. During migration it helps to set everything to private and only make public the functions that are accessed by external classes. 

In this case, I discovered that only two fields of the Parser class were ever accessed by external classes:

```
public parse(line: string): Result { ... }

public static DEBUG: boolean = false;
```

However, I'm happy to remove the public / private modifiers to be consistent with the rest of the code. I will keep instance variables protected as you mentioned earlier.